File: server/src/schema/movement.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MovementActionRequestV1 {
    Dash,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MovementActionV1 {
    None,
    Dashing,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum MovementZoneKindV1 {
    Normal,
    Dead,
    Negative,
    ResidueAsh,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct MovementStateV1 {
    pub current_speed_multiplier: f32,
    pub stamina_cost_active: bool,
    pub movement_action: MovementActionV1,
    pub zone_kind: MovementZoneKindV1,
    pub dash_cooldown_remaining_ticks: u64,
    pub hitbox_height_blocks: f32,
    pub stamina_current: f32,
    pub stamina_max: f32,
    pub low_stamina: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub last_action_tick: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rejected_action: Option<MovementActionRequestV1>,
}

File: agent/packages/schema/src/movement.ts
import { Type, type Static } from "@sinclair/typebox";

export const MovementActionRequestV1 = Type.Union([
  Type.Literal("dash"),
  Type.Literal("slide"),
  Type.Literal("double_jump"),
]);
export type MovementActionRequestV1 = Static<typeof MovementActionRequestV1>;

export const MovementActionV1 = Type.Union([
  Type.Literal("none"),
  Type.Literal("dashing"),
  Type.Literal("sliding"),
  Type.Literal("double_jumping"),
]);
export type MovementActionV1 = Static<typeof MovementActionV1>;

export const MovementZoneKindV1 = Type.Union([
  Type.Literal("normal"),
  Type.Literal("dead"),
  Type.Literal("negative"),
  Type.Literal("residue_ash"),
]);
export type MovementZoneKindV1 = Static<typeof MovementZoneKindV1>;

export const MovementStateV1 = Type.Object(
  {
    current_speed_multiplier: Type.Number({ minimum: 0 }),
    stamina_cost_active: Type.Boolean(),
    movement_action: MovementActionV1,
    zone_kind: MovementZoneKindV1,
    dash_cooldown_remaining_ticks: Type.Integer({ minimum: 0 }),
    slide_cooldown_remaining_ticks: Type.Integer({ minimum: 0 }),
    double_jump_charges_remaining: Type.Integer({ minimum: 0 }),
    double_jump_charges_max: Type.Integer({ minimum: 0 }),
    hitbox_height_blocks: Type.Number({ minimum: 0 }),
    stamina_current: Type.Number({ minimum: 0 }),
    stamina_max: Type.Number({ exclusiveMinimum: 0 }),
    low_stamina: Type.Boolean(),
    last_action_tick: Type.Optional(Type.Integer({ minimum: 0 })),
    rejected_action: Type.Optional(MovementActionRequestV1),
  },
  { additionalProperties: false },
);
export type MovementStateV1 = Static<typeof MovementStateV1>;
